server/actions/MongoDriver.ts

```diff
@@ -0,0 +1,104 @@
+import mongoDb from '../index'
+import { Document, FilterQuery, Model } from 'mongoose'
+import '../../utils/types'
+import { ApiError, ItemDefinition, ServerModel } from '../../utils/types'
+
+const ENTITY_NOT_FOUND_MESSAGE = 'Entity does not exist'
+
+/**
+ * Gets a single Entity object from the database with the given id
+ * @param dbSchema - The schema of the entity object to get
+ * @param id - The _id of the entity object to get
+ * @returns A single entity object
+ */
+export async function getEntity<Schema extends Document>(
+  dbSchema: Model<Schema>,
+  id: string
+) {
+  await mongoDb()
+
+  let response = await dbSchema.findById(id)
+  if (!response) throw new ApiError(404, ENTITY_NOT_FOUND_MESSAGE)
+  return response
+}
+
+/**
+ * Returns all entities from a collection
+ * @param dbSchema The collection schema to get entities from
+ * @returns A list of all entities in the collection
+ */
+export async function getEntities<Schema extends Document>(
+  dbSchema: Model<Schema>
+) {
+  await mongoDb()
+
+  let response = await dbSchema.find()
+  return response
+}
+
+/**
+ * Creates a new Entity object in the database
+ * @param dbSchema The collection schema to get entities from
+ * @param document - The Entity object to create
+ * @returns The newly created Entity object in the database
+ */
+export async function createEntity<
+  Schema extends Document,
+  T extends ServerModel
+>(dbSchema: Model<Schema>, document: T) {
+  await mongoDb()
+
+  let response = await dbSchema.create(document)
+  return response
+}
+
+/**
+ * Replaces the existing Entity object with the given id with a new entity
+ * @param dbSchema The collection schema to get entities from
+ * @param id - _id of the Entity object to update
+ * @param document - The new Entity object to replace the existing Entity object with
+ */
+export async function updateEntity<
+  Schema extends Document,
+  T extends ServerModel
+>(dbSchema: Model<Schema>, id: string, document: T) {
+  await mongoDb()
+
+  let response = await dbSchema.findByIdAndUpdate(id, document)
+  if (!response) throw new ApiError(404, ENTITY_NOT_FOUND_MESSAGE)
+}
+
+/**
+ * Deletes the Entity object with the given id
+ * @param dbSchema The collection schema to get entities from
+ * @param id - The _id of the Entity object to delete
+ */
+export async function deleteEntity<Schema extends Document>(
+  dbSchema: Model<Schema>,
+  id: string
+) {
+  await mongoDb()
+
+  await dbSchema.findByIdAndDelete(id, (response: any) => {
+    if (!response) {
+      throw new ApiError(404, ENTITY_NOT_FOUND_MESSAGE)
+    }
+  })
+}
+
+export async function findEntities<
+  Schema extends Document,
```

Comment:
I was playing around with the type of this a bit and I think we can use TypeScript to ensure that the schema passed in lines up with the `filterDocument` passed. check this out and let me know what you think:

```ts
export async function findEntities<
  T extends ServerModel,
  ModelDocument extends Omit<T, '_id'> & Document
>(dbSchema: Model<ModelDocument>, filterDocument: Partial<T> | T
```

Reply:
fixed. nice!